File: Src/PythonServer/agent_framwork/bindable_property/bindable_property.py
import asyncio
from typing import Any, Callable, List, Union
import inspect
import threading


class IUnRegister:
    def un_register(self):
        raise NotImplementedError()


class BindableProperty:
    def __init__(self, initial_value: Any = None):
        self._value = initial_value
        self._callbacks: List[Callable[[Any], None]] = []

        # 根据是否在异步环境下决定使用哪种锁
        try:
            loop = asyncio.get_event_loop()
            self._lock = asyncio.Lock() if loop.is_running() else threading.Lock()
        except RuntimeError:
            self._lock = threading.Lock()

    @property
    def value(self) -> Any:
        return self._value

    async def aget_value(self) -> Any:
        """异步获取值"""
        return self.value

    async def aset_value(self, new_value: Any):
        """异步设置值"""
        if inspect.isawaitable(self._lock):
            async with self._lock:
                await self._set_value(new_value)
        else:
            with self._lock:
                await self._set_value(new_value)

    def set_value(self, new_value: Any):
        """同步设置值"""
        if inspect.isawaitable(self._lock):
            asyncio.run(self.aset_value(new_value))
        else:
            with self._lock:
                self._set_value_sync(new_value)

    def _set_value_sync(self, new_value: Any):
        if new_value != self._value:
            self._value = new_value
            for callback in self._callbacks:
                self._invoke_callback(callback, new_value)

    async def _set_value(self, new_value: Any):
        if new_value != self._value:
            self._value = new_value
            for callback in self._callbacks:
                self._invoke_callback(callback, new_value)

    def _invoke_callback(self, callback: Callable[[Any], None], value: Any):
        if inspect.iscoroutinefunction(callback):
            asyncio.create_task(callback(value))
        else:
            callback(value)

    def register_on_value_changed(self, callback: Callable[[Any], None]) -> IUnRegister:
        self._callbacks.append(callback)
        return BindablePropertyUnRegister(bindable_property=self, callback=callback)

    def unregister_on_value_changed(self, callback: Callable[[Any], None]):
        if callback in self._callbacks:
            self._callbacks.remove(callback)


class BindablePropertyUnRegister(IUnRegister):
    def __init__(self, bindable_property: BindableProperty, callback: Callable[[Any], None]):
        self.bindable_property = bindable_property
        self.callback = callback

    def un_register(self):
        if self.bindable_property and self.callback:
            self.bindable_property.unregister_on_value_changed(self.callback)
        self.bindable_property = None
        self.callback = None

File: Src/IndependentAgentProject/Assets/FrameworkDesign/Framework/BindableProperty/BindableProperty.cs
using System;

namespace FrameworkDesign
{
    public class BindableProperty<T>
    {
        public T mValue = default(T);//比如int为0、bool为false等
        public T Value
        {
            get => mValue;
            set
            {
                if(!value.Equals(mValue))
                {
                    mValue = value;

                    //OnValueChanged?.Invoke(Value);
                    mOnValueChanged?.Invoke(Value);
                }
            }
        }
        //public Action<T> OnValueChanged;
        private Action<T> mOnValueChanged = (v) => { };
        public IUnRegister RegisterOnValueChanged(Action<T> onValueChanged)
        {
            mOnValueChanged = onValueChanged;
            return new BindablePropertyUnRegister<T>()
            {
                bindableProperty = this,
                OnValueChanged = onValueChanged
            };
        }
        public void UnRegisterOnValueChanged(Action<T> onValueChanged)
        {
            mOnValueChanged -= onValueChanged;
        }
    }
    //IUnRegister是TypeEventSystem部分里，返回的用于注销的类
    //记录了bindableProperty、其上注册的OnValueChanged方法，用于把OnValueChanged从bindableProperty注销掉
    public class BindablePropertyUnRegister<T> : IUnRegister
    {
        public BindableProperty<T> bindableProperty { get; set; }
        public Action<T> OnValueChanged { get; set; }
        public void UnRegister()
        {
            bindableProperty.UnRegisterOnValueChanged(OnValueChanged);
            bindableProperty = null;
            OnValueChanged = null;
        }
    }
}


